apps/src/gamelab/GameLab.js

```diff
@@ -471,6 +483,84 @@ GameLab.prototype.onArrowButtonUp = function (buttonId, e) {
   this.gameLabP5.notifyKeyCodeUp(p5KeyCodeFromArrow(buttonId));
 };
 
+GameLab.prototype.onDPadButtonDown = function (e) {
+  this.dPadState = {};
+  this.dPadState.boundHandler = this.onDPadMouseMove.bind(this);
+  document.body.addEventListener('mousemove', this.dPadState.boundHandler);
+  this.dPadState.touchEventName = dom.getTouchEventName('mousemove');
+  if (this.dPadState.touchEventName) {
+    document.body.addEventListener(this.dPadState.touchEventName,
+        this.dPadState.boundHandler);
+  }
+  this.dPadState.startingX = e.clientX;
+  this.dPadState.startingY = e.clientY;
+  this.dPadState.previousX = e.clientX;
+  this.dPadState.previousY = e.clientY;
+
+  $('#studio-dpad-button').addClass('active');
+
+  e.preventDefault();  // Stop normal events so we see mouseup later.
+};
+
+var DPAD_DEAD_ZONE = 3;
+
+GameLab.prototype.onDPadMouseMove = function (e) {
+  var dPadButton = $('#studio-dpad-button');
+  var self = this;
+
+  function notifyKeyHelper(keyCode, cssClass, start, prev, cur, invert) {
+    if (invert) {
+      start *= -1;
+      prev *= -1;
+      cur *= -1;
+    }
+    start -= DPAD_DEAD_ZONE;
+
+    if (cur < start) {
+      if (prev >= start) {
+        self.gameLabP5.notifyKeyCodeDown(keyCode);
+        dPadButton.addClass(cssClass);
+      }
+    } else if (prev < start) {
+      self.gameLabP5.notifyKeyCodeUp(keyCode);
+      dPadButton.removeClass(cssClass);
```

Comment:
If the `prev` values are updated after each call, it should be fine as far as I can tell. If not, then you'd definitely see it would definitely do work over and over. Which isn't particularly dangerous, but not very efficient to send extra events.